syntax = "proto3";
package csce438;
import "google/protobuf/timestamp.proto";

// ------------------------------------------------------------
// The coordinator service definition
// These are the different functionalities exposed by service
// ------------------------------------------------------------

//Init and Heartbeat potentially redundant
service CoordService{
  rpc Heartbeat (ServerInfo) returns (Confirmation) {}
  rpc GetServer (ID) returns (ServerInfo) {}
  rpc CreatePath (ServerInfo) returns (Confirmation) {}
  //rpc exists (Path) (Confirmation) {}
  // ZooKeeper API here
  rpc RegisterSyncServer(ServerInfo) returns (Confirmation){}
  rpc GetAllFollowerServers (ID) returns (ServerList) {}
  rpc GetFollowerServer (ID) returns (ServerInfo) {}
  rpc GetCounterpart (ServerInfo) returns (ServerInfo) {}
}

//server info message definition
message ServerInfo{
  int32 serverID = 1;
  int32 clusterId = 2;
  string hostname = 3;
  string port = 4;
  string type = 5;
}

//message Register{
//  ServerInfo serverInfo = 1;
//  int32 clusterId = 2;
//}


//confirmation message definition
message Confirmation{
  bool status = 1;
  bool is_master = 2;
}

//id message definition
message ID{
  int32 id = 1;
}

message ServerList{
  repeated int32 serverID = 1;
  repeated string hostname = 2;
  repeated string port = 3;
  repeated string type = 4;
}


//TODO:
//potentially add a wake up synchronization rpc to be made from
//coordinator when it detects a server reconnects after being off
service SynchService{
  rpc GetAllUsers (Confirmation) returns (AllUsers) {}
  rpc GetTLFL (ID) returns (TLFL) {}
  rpc ResynchServer (ServerInfo) returns (Confirmation) {}
  rpc SynchFollower (FollowRequest) returns (Confirmation) {}
}

//list of all users known of by given follower synchronizer
message AllUsers{
  repeated string users = 1;
}

message FollowRequest{
  int32 user1 = 1;
  int32 user2 = 2;
}

//two repeated string lists of each entry in a users 
//Timeline (TL) and Follower List (FL)
message TLFL{
  bool status = 1;
  repeated string TL = 2;
  repeated string FL = 3;
}


